students/khtruong/lesson_09/mailroom_oop.py

```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env python
+
+"""This module runs the mailroom script using oop paradigm."""
+
+from datetime import date
+
+
+class Donor:
+    def __init__(self, firstname, lastname, donations=None):
+        self.firstname = firstname
+        self.lastname = lastname
+        self.fullname = '{} {}'.format(firstname, lastname)
+        if donations:
+            self.donations = donations
+        else:
+            self.donations = []
+
+    def __lt__(self, other):
+        return ((self.totaldonation, self.firstname) <
+                (other.totaldonation, other.firstname))
+
+    def add_donation(self, amount):
+        self.donations.append(amount)
+
+    @property
+    def totaldonation(self):
+        return sum(self.donations)
+
+    @property
+    def numdonation(self):
+        return len(self.donations)
+
+    @property
+    def avgdonation(self):
+        return self.totaldonation / max(self.numdonation, 1)
+
+    @classmethod
+    def from_fullname(cls, fullname, donations):
+        firstname, lastname = fullname.split(' ')
+        return cls(firstname, lastname, donations)
+
+
+class DataBase:
+
+    def __init__(self, donors=None):
+        if donors:
+            self.donors = donors
+        else:
+            self.donors = []
+
+    def donor_and_amount(self):
+        """Return prompt for donor name and donation amount and print to
+        screen.
+        """
+        fullname = fullname_input()
+        amount = amount_input()
+        for donor in self.donors:
+            if fullname == donor.fullname:
+                donor.add_donation(amount)
+                break
+        else:
+            self.add_donor(fullname, [amount])
+        print(self._format_letter(fullname, amount))
+
+    def add_donor(self, fullname, donations=None):
+        self.donors.append(Donor.from_fullname(fullname, donations))
+
+    def list_donors(self):
+        """Return all donors name."""
+        for donor in self.donors:
+            print(donor.fullname)
+
+    def create_report(self):
+        """Return a summary report of donations."""
+        header = '\nDonor Name                | Total Given |' \
+                 ' Num Gifts | Average Gift'
+        print(header)
+        print('-' * (len(header) - 1))
+        sortedlist = sorted(self.donors, reverse=True)
+        for donor in sortedlist:
+            print('{:<26} ${:>11,.2f} {:>11d}  ${:>12,.2f}'.format(
+                donor.fullname, donor.totaldonation,
+                donor.numdonation, donor.avgdonation))
+
+    def send_letters(self):
+        """Create thank you letter for each donor and save as text files"""
+        for donor in self.donors:
+            fullname = donor.fullname
+            amount = donor.donations[-1]
+            filename = ''.join([fullname, ' ',
+                                str(date.today()),
+                                '.txt'])
```

Comment:
something like an f-string would read better here:
`f"{fullname}_{date.today()}.txt"` (I would also avoid empty spaces in file names)